libraries/AP_Telemetry/AP_Telemetry_MQTT.cpp

```diff
@@ -0,0 +1,252 @@
+        telemetry_mqtt->init_mqtt();
+    }
+    return telemetry_mqtt;
+}
+
+AP_Telemetry_MQTT::AP_Telemetry_MQTT(AP_Telemetry &frontend, AP_HAL::UARTDriver* uart) :
+    AP_Telemetry_Backend(frontend, uart)
+{}
+
+void AP_Telemetry_MQTT::init_mqtt()
+{
+    int rc;
+    mqtt_mutex_store = PTHREAD_MUTEX_INITIALIZER;
+    mqtt_mutex = &mqtt_mutex_store;
+    pthread_mutexattr_t attr;
+    pthread_mutexattr_init(&attr);
+
+    recv_msg_list = ListInitialize();
+    pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_ERRORCHECK);
+    if ((rc = pthread_mutex_init(mqtt_mutex, &attr)) != 0) {
+        printf("init: error %d initializing mqtt_mutex\n", rc);
+        MQTTHandle_error(rc);
+    } else {
+
+        conn_options.keepAliveInterval = 20;
+        conn_options.cleansession = 1;
+
+        conn_options.onSuccess = onConnect;
+        conn_options.onFailure = onConnectFailure;
+        conn_options.context = mqtt_client;
+
+        char clientid[9] = "no_id";
+        srand((unsigned int)time(0));
+        sprintf(clientid, "client_%d", rand()%100);
+
+        if ((rc = MQTTAsync_create(&mqtt_client, mqtt_server, clientid, MQTTCLIENT_PERSISTENCE_NONE, NULL)) != 0) {
+            printf("Failed to create Client, return code %d\n", rc);
+            MQTTHandle_error(rc);
+        }
+
+        MQTTAsync_setCallbacks(mqtt_client, NULL, NULL, mqtt_msg_arrived, NULL);
+        if ((rc = MQTTAsync_connect(mqtt_client, &conn_options)) != MQTTASYNC_SUCCESS) {
+            printf("Failed to start connect, return code %d\n", rc);
+            MQTTHandle_error(rc);
+        }
+        connection_status = MQTT_CONNECTED;
+    }
+}
+
+void AP_Telemetry_MQTT::set_mqtt_server(const char* server)
+{
+    mqtt_server = server;
+}
+
+void AP_Telemetry_MQTT::set_mqtt_user(const char* user)
+{
+    conn_options.username = user;
+}
+
+void AP_Telemetry_MQTT::set_mqtt_password(const char* password)
+{
+    conn_options.password = password;
+}
+
+void AP_Telemetry_MQTT::send_log(const char* str)
+{
+    char topic[MAX_PAYLOAD];
+    if (send_log_flag == MQTT_SEND_LOG_ON) {
+        if (connection_status == MQTT_CONNECTED) {
+            sprintf(topic, "$ardupilot/copter/quad/log/%04d/location",
+                    mavlink_system.sysid);
+            send_message(str, topic);
+        }
+    }
+}
+
+void AP_Telemetry_MQTT::send_message(const char *str, const char *topic)
+{
+    int rc;
+    MQTTAsync_message pubmsg = MQTTAsync_message_initializer;
+    char tmp[MAX_PAYLOAD];
+
+    strcpy(tmp, str);
+    pubmsg.payload = tmp;
+    pubmsg.payloadlen = strlen(str);
+    pubmsg.qos = QOS;
+    pubmsg.retained = 0;
+    if ((rc = MQTTAsync_sendMessage(mqtt_client, topic, &pubmsg, NULL)) != MQTTASYNC_SUCCESS) {
+        printf("Failed to start sendMessage, return code %d\n", rc);
+        MQTTHandle_error(rc);
+    }
+}
+
+void AP_Telemetry_MQTT::subscribe_mqtt_topic(const char *topic, int qos)
+{
+    if (connection_status == MQTT_CONNECTED) {
+        int rc;
```

Comment:
RC is an abbreviation for radio control. I think that it is better not to overlap with major abbreviations.